File: solutions/0202_happy_number/python/solution_fast-slow-pointers.py
class Solution:
    def isHappy(self, num: int) -> bool:
        def calculate_square_sum(num):
            res = 0

            while num:
                digit = num % 10
                res = res + digit * digit
                num = num // 10

            return res

        slow = calculate_square_sum(num)
        fast = calculate_square_sum(calculate_square_sum(num))

        while slow != fast:
            slow = calculate_square_sum(slow)
            fast = calculate_square_sum(calculate_square_sum(fast))

        return slow == 1

File: solutions/0202_happy_number/cpp/solution_fast-slow-pointers.cpp
class Solution {
 public:
  bool isHappy(int num) {
    int slow = sumSquare(num);
    int fast = sumSquare(sumSquare(num));

    while (slow != fast) {
      slow = sumSquare(slow);
      fast = sumSquare(sumSquare(fast));
    }

    return slow == 1;
  }

 private:
  int sumSquare(int num) {
    int sum = 0;

    while (num) {
      sum += pow(num % 10, 2);
      num /= 10;
    }

    return sum;
  }
};